logstash/class-logger.php

```diff
@@ -0,0 +1,439 @@
+
+		} elseif ( isset( $data['user_ua'] ) && ! is_string( $data['user_ua'] ) ) {
+			trigger_error( 'Invalid `user_ua` in call to ' . esc_html( __METHOD__ ) . '(). Must be a string.', E_USER_WARNING );
+			return; // Failed validation.
+
+		} elseif ( isset( $data['user_ua'] ) && strlen( $data['user_ua'] ) > static::MAX_ENTRY_USER_UA_SIZE ) {
+			trigger_error( 'Invalid `user_ua` in call to ' . esc_html( __METHOD__ ) . '(). Must be ' . esc_html( static::MAX_ENTRY_USER_UA_SIZE ) . ' bytes or less.', E_USER_WARNING );
+			return; // Failed validation.
+
+		} elseif ( isset( $data['extra'] ) && ! is_array( $data['extra'] ) && ! is_object( $data['extra'] ) && ! is_scalar( $data['extra'] ) ) {
+			trigger_error( 'Invalid `extra` in call to ' . esc_html( __METHOD__ ) . '(). Must be an object, array, or scalar value.', E_USER_WARNING );
+			return; // Failed validation.
+
+		} elseif ( isset( $data['extra'] ) && strlen( wp_json_encode( $data['extra'] ) ) > static::MAX_ENTRY_EXTRA_SIZE ) {
+			trigger_error( 'Invalid `extra` in call to ' . esc_html( __METHOD__ ) . '(). Must be ' . esc_html( static::MAX_ENTRY_EXTRA_SIZE ) . ' bytes or less.', E_USER_WARNING );
+			return; // Failed validation.
+		}
+
+		// Adjust `feature` and JSON-encode pretty-print data.
+		$data['feature'] = 'a8c_vip_' . $data['feature'];
+
+		// Adjust `extra` and JSON-encode data that gets sent in the API request.
+		$data['extra'] = isset( $data['extra'] ) ? wp_json_encode( $data['extra'], JSON_PRETTY_PRINT ) : '';
+
+		// Adds a new log entry.
+		static::$entries[] = $data;
+
+		// Sends data to logstash on shutdown.
+		if ( ! has_action( 'shutdown', [ static::class, 'process_entries_on_shutdown' ] ) ) {
+			add_action( 'shutdown', [ static::class, 'process_entries_on_shutdown' ], 9999 );
+		}
+	}
+
+	/**
+	 * Sends data to logstash via REST API on shutdown, viewable in Kibana.
+	 *
+	 * @since 2020-01-10
+	 */
+	public static function process_entries_on_shutdown() : void {
+		if ( static::$processed_entries ) {
+			return; // Already done.
+		}
+
+		$fallback_error = new \WP_Error( 'logstash-send-failed', 'There was an error connecting to the logstash endpoint' );
+
+		static::$processed_entries = true;
+
+		if ( function_exists( 'fastcgi_finish_request' ) ) {
+			// Flush content to client first to prevent slow page load
+			fastcgi_finish_request();
+		}
+
+		$endpoint = 'https://public-api.wordpress.com/rest/v1.1/logstash/bulk';
+
+		$entry_chunks = array_chunk( static::$entries, static::BULK_ENTRIES_COUNT );
+
+		// Process all entries.
+		foreach ( $entry_chunks as $entries ) {
+			if ( ! defined( 'VIP_GO_ENV' ) || ! VIP_GO_ENV ) {
+				static::maybe_wp_debug_log_entries( $entries );
+				continue; // Bypassing REST API below in this case.
+			}
+
+			$json_data = wp_json_encode( $entries );
+
+			// Send to logstash via REST API endpoint with payload containing log entry details.
+			$_wp_remote_response = vip_safe_wp_remote_request( $endpoint, $fallback_error, 3, 2, 5, [
```

Comment:
Now that we switched to using `fastcgi_finish_request()`, we can bump up the timeouts here since this won't be blocking the client request. ~10 seconds is probably ok here.

Reply:
Good point but `vip_safe_wp_remote_request()` maxes out the timeout value at 5 seconds

Reply:
Changed in 33fcad1bcf9ca21ee9ef1772481d14b09c8eb31d

Reply:
Oh right - ok 5 should be fine...